### DISCUSSION TITLE
url request missing parameters when using serverless-pre-function plugin

### DISCUSSION THREAD

**Discussion by engimatic** [apache/apisix#8216] [Q&A]
### Description

插件代码
`return function(conf, ctx) local core = require("apisix.core");local http =
    require("resty.http");local httpc = http.new(); local cjson = require
    "cjson";local cjson2 = cjson.new();local authorization =
    core.request.header(ctx, "Authorization"); local uri =
    ctx.var.uri;if(authorization == nil) then ngx.log(ngx.ERR, "match uri ",
    uri);core.response.exit(401);return; end;local url =
    "http://ip:port/checkToken";local
    req = {};req.uri = uri;req.token = authorization; local res,err =
    httpc:request_uri(url, {method = "POST", body = cjson2.encode(req), headers
    = {
        ["Content-Type"] = "application/json"
    }});if not res then  core.response.exit(500);return;end;local result =
    cjson2.decode(res.body); if not(result["data"]) then ngx.log(ngx.ERR,
    "RESULT ", res.body);core.response.exit(res.status, res.body);return;end;end`

调用接口/v1/sys/group/userLike?name=aa，连续多次请求，后端发现部分请求url没有name参数，name参数被截断。禁用插件后，没有问题

### Environment

- APISIX version :2.15.0
- Operating system :k8s

**Comment by tokers** [comment]
@engimatic Is this your code? Try to troubleshoot it.

**Comment by engimatic** [comment]
> @engimatic Is this your code? Try to troubleshoot it.

@tokers 
![image](https://user-images.githubusercontent.com/14733591/198914132-609edb3c-bdb3-4964-a16b-5c6b509b8eea.png)

这是我的erverless-pre-function插件代码。 /v1/sys/group/userLike?name=aa。每100次请求，可能出现4-5个参数被截断的问题。post请求json格式的request，都没有出现过问题

**Comment by engimatic** [comment]
@tokers 这个是我们做鉴权的自定义插件，目前插件代码运行没有任何问题

**Comment by tzssangglass** [comment]
> 调用接口/v1/sys/group/userLike?name=aa，连续多次请求，后端发现部分请求url没有name参数，name参数被截断。禁用插件后，没有问题

can you describe this problem clearly?

**Reply by engimatic** [reply]
> > 调用接口/v1/sys/group/userLike?name=aa，连续多次请求，后端发现部分请求url没有name参数，name参数被截断。禁用插件后，没有问题
> 
> can you describe this problem clearly?

我有一个接口，GET HTTPS://IP:PORT/v1/sys/group/userLike?name=aa。
接口通过apisix路由到后端服务。我们启用的插件如上所述，代码一摸一样。
现在每一百个请求，会有4-5个报错“Required request parameter 'name' for method parameter type String is not present”。不使用上述插件中的http请求部分，就不会出现错误

**Reply by tzssangglass** [reply]
> 我有一个接口，GET [HTTPS://IP:PORT/v1/sys/group/userLike?name=aa。](HTTPS://IP:PORT/v1/sys/group/userLike?name=aa%E3%80%82)

`name=aa` is uri args, from your code, I don't find the code to get the uri args.

**Reply by engimatic** [reply]
> > 我有一个接口，GET [HTTPS://IP:PORT/v1/sys/group/userLike?name=aa。](HTTPS://IP:PORT/v1/sys/group/userLike?name=aa%E3%80%82)
> 
> `name=aa` is uri args, from your code, I don't find the code to get the uri args.


write with springboot
`@ApiOperation(value = "人员模糊查询")
    @GetMapping("/userLike")
    public Result<List<GroupUserMixedVo>> userLike(@RequestParam(value = "name") String name) {
        return Result.ok(userApp.userLike(name));
    }`

**Reply by tzssangglass** [reply]
> write with springboot
> `@ApiOperation(value = "人员模糊查询") @GetMapping("/userLike") public Result<List<GroupUserMixedVo>> userLike(@RequestParam(value = "name") String name) { return Result.ok(userApp.userLike(name)); }`

What I mean is that in the lua code you don't get the uri agrs and send it to spring boot via the resty-http.

`ngx.var.uri` only get `/v1/sys/group/userLike`, not `/v1/sys/group/userLike?name=Apache APISIX`

**Reply by engimatic** [reply]
> > write with springboot
> > `@ApiOperation(value = "人员模糊查询") @GetMapping("/userLike") public Result<List<GroupUserMixedVo>> userLike(@RequestParam(value = "name") String name) { return Result.ok(userApp.userLike(name)); }`
> 
> What I mean is that in the lua code you don't get the uri agrs and send it to spring boot via the resty-http.
> 
> `ngx.var.uri` only get `/v1/sys/group/userLike`, not `/v1/sys/group/userLike?name=Apache APISIX`

checkAuth api just check whether this user  has authorization of API  `/v1/sys/group/userLike`,not real call the API

**Reply by tzssangglass** [reply]
> checkAuth api just check whether this user has authorization of API `/v1/sys/group/userLike`,not real call the API

I review your plugin code, it does not modify `ctx.var.upstream_uri`. there is no stable reproduction of the steps I can not provide more help.

**Reply by engimatic** [reply]
> > checkAuth api just check whether this user has authorization of API `/v1/sys/group/userLike`,not real call the API
> 
> I review your plugin code, it does not modify `ctx.var.upstream_uri`. there is no stable reproduction of the steps I can not provide more help.

1.  start a springboot application with API like this

`
 @GetMapping("/userLike")
    public String userLike(@RequestParam(value = "name") String name) {
        return name;
    }

@PostMapping("/checkToken")
    public Result<Boolean> checkToken(HttpServletResponse response, @RequestBody CheckTokenReq checkTokenReq) throws IOException, AppNotFoundException {
        Result.ok(true);
    }
`

3.  use APISIX with plugin like this : use serverless-pre-function plugin

`
return function(conf, ctx) local core = require("apisix.core");local http = require("resty.http");local httpc = http.new(); local cjson = require "cjson";local cjson2 = cjson.new();local authorization = core.request.header(ctx, "Authorization"); local uri = ctx.var.uri;if(authorization == nil) then ngx.log(ngx.ERR, "match uri ", uri);core.response.exit(401);return; end;local url = "http://ip:port/checkToken";local req = {};req.uri = uri;req.token = authorization; local res,err = httpc:request_uri(url, {method = "POST", body = cjson2.encode(req), headers = { ["Content-Type"] = "application/json" }});if not res then core.response.exit(500);return;end;local result = cjson2.decode(res.body); if not(result["data"]) then ngx.log(ngx.ERR, "RESULT ", res.body);core.response.exit(res.status, res.body);return;end;end
`

5. call this 1000 times with postman.
You would found something wrong  in the 1000 times  call。

**Reply by tzssangglass** [reply]
I can't reproduce

1. java code

```java
package com.example.mock;

import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PatchMapping;
import org.springframework.web.bind.annotation.RequestParam;

import org.springframework.web.bind.annotation.RestController;

@RestController
public class MockController {

    @GetMapping("/userLike")
    public String userLike(@RequestParam(value = "name") String name) {
        return name;
    }

    @PatchMapping ("/checkToken")
    public Boolean checkToken() {
        return true;
    }
}
```

2. route

```
curl -i http://127.0.0.1:9180/apisix/admin/routes/2  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' -X PUT -d '
{
    "uri": "/user",
    "plugins": {
        "serverless-pre-function": {
            "phase": "rewrite",
            "functions" : ["return function(conf, ctx) local core = require(\"apisix.core\");local http = require(\"resty.http\");local httpc = http.new(); local cjson = require \"cjson\";local cjson2 = cjson.new();local authorization = core.request.header(ctx, \"Authorization\"); local uri = ctx.var.uri;if(authorization == nil) then ngx.log(ngx.ERR, \"match uri \", uri);core.response.exit(401);return; end;local url = \"http://127.0.0.1:8080/checkToken\";local req = {};req.uri = uri;req.token = authorization; local res,err = httpc:request_uri(url, {method = \"POST\", body = cjson2.encode(req), headers = { [\"Content-Type\"] = \"application/json\" }});if not res then core.response.exit(500);return;end;local result = cjson2.decode(res.body); end"]
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "127.0.0.1:8080": 1
        }
    }
}'
```

3. test shell

```
#!/usr/bin/env bash

for i in {1..1}; do \
    name=$(curl -s "http://127.0.0.1:9080/user?name=bar" -H "Authorization: fake")
    if [ $name != "bar" ]; then
        echo "failed to get name: " $name
        exit 1
    fi
done

echo "success"
```

I get `success` after execute `test.sh`

**Reply by engimatic** [reply]
> for i in {1..1}

for i in {1..1}   you just test once.In my case, I try 5000 times, it appear 4-5 times.You can change to `for i in {1..5000}`

**Reply by tzssangglass** [reply]
In fact, I tried 3*1000 times without exception.

**Reply by engimatic** [reply]
> In fact, I tried 3*1000 times without exception.

I have tried again many times.When my backend service deploy in k8s, it happend again.When my backend service deploy in virtual machine，there is no errors.So it's maybe caused by k8s network, I think.But I cannot find what wrong with it.

`
web-> k8s LB -> k8s apisix -> k8s backend, error appear sometimes;  

web-> k8s LB -> k8s apisix -> virtual machine backend, no errors.
`

**Reply by tzssangglass** [reply]
> I have tried again many times.When my backend service deploy in k8s, it happend again.When my backend service deploy in virtual machine，there is no errors.So it's maybe caused by k8s network, I think.But I cannot find what wrong with it.

As bizarre as it sounds, I haven't had a similar problem. But from experience, it could be that there is a configuration error somewhere, replacing the original path that contain uri args with the request path

**Reply by engimatic** [reply]
> > I have tried again many times.When my backend service deploy in k8s, it happend again.When my backend service deploy in virtual machine，there is no errors.So it's maybe caused by k8s network, I think.But I cannot find what wrong with it.
> 
> As bizarre as it sounds, I haven't had a similar problem. But from experience, it could be that there is a configuration error somewhere, replacing the original path that contain uri args with the request path

But it not happend everytime,just a certain probability occur.

**Reply by engimatic** [reply]
> > I have tried again many times.When my backend service deploy in k8s, it happend again.When my backend service deploy in virtual machine，there is no errors.So it's maybe caused by k8s network, I think.But I cannot find what wrong with it.
> 
> As bizarre as it sounds, I haven't had a similar problem. But from experience, it could be that there is a configuration error somewhere, replacing the original path that contain uri args with the request path

I have test another case:
K8S LB -> K8S APISIX -> K8S SERVICE     error appear sometimes;
K8S LB -> K8S APISIX -> K8S SERVICE,no plugin; no errors;
K8S LB -> K8S APISIX -> virtual machine SERVICE    no errors;
K8S LB -> K8S SERVICE   no errors;

I  was utterly confused.

**Reply by engimatic** [reply]
I have just find something wrong with tomcat and i had test more and more,find that in other case, there are errors with small probability.I think i have reloved this problem.
THANKS !
